linera-views/src/shared_view.rs

```diff
@@ -0,0 +1,220 @@
+// Copyright (c) Zefchain Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+#[cfg(test)]
+#[path = "unit_tests/shared_view.rs"]
+mod tests;
+
+use crate::{
+    batch::Batch,
+    common::Context,
+    views::{RootView, View, ViewError},
+};
+use async_lock::{Semaphore, SemaphoreGuardArc};
+use async_trait::async_trait;
+use futures::channel::oneshot;
+use std::{
+    marker::PhantomData,
+    ops::{Deref, DerefMut},
+    sync::{
+        atomic::{AtomicUsize, Ordering},
+        Arc, Mutex,
+    },
+};
+
+#[cfg(not(target_arch = "wasm32"))]
+use crate::{increment_counter, SAVE_VIEW_COUNTER};
+
+/// A way to safely share a [`View`] among multiple readers and at most one writer.
+///
+/// [`View`]s represent some data persisted in storage, but it also contains some state in
+/// memory that caches the storage state and that queues changes to the persisted state to
+/// be sent later. This means that two views referencing the same data in storage may have
+/// state conflicts in memory, and that's why they can't be trivially shared (using
+/// [`Clone`] for example).
+///
+/// The [`SharedView`] provides a way to share an inner [`View`] more safely, by ensuring
```

Comment:
The expression "more safely" does not mean "safe". I note that the documentation never says that the state of the views are always valid.

Reply:
>  I note that the documentation never says that the state of the views are always valid.

Nothing prevents us from documenting the reasonable working assumptions under which data validity is guaranteed.